src/toplc.ml

```diff
@@ -656,17 +651,92 @@ let get_ancestors h c =
+  List.concat
+  [ 
+      List.concat (List.mapi (fun i param -> bc_store (max - i) param) rev_params )
+    ; List.concat (List.mapi (fun i param -> bc_load (locals + 1 + i) param) parameters)
+    ; bc_emit call_id (tag_from (locals+1) parameters) ] in
+  let instrumentation = put_labels_on instrumentation in
+  instrumentation @ instrs
+
+let instrument_return locals return return_id instrs =
+  let instrumentation =
+    if log_in then printf "Return Instrumented with return ID: %a\n"
+                          (U.pp_option U.pp_int) return_id ;
+  if_ (has return_id)
+      [ if_ (return <> `Void)
+      [ bc_dup return
+      ; bc_store locals return ]
+    ; bc_emit return_id (if_ (return <> `Void) [[(locals, return)]]) ]
+  in
+  let instrumentation = put_labels_on instrumentation in
+  instrs @ instrumentation
+
+let instrument_invoke method_name class_name parameters return env locals instr =
+     let method_name = string_of_method_name method_name in
+     let full_method_name = mk_full_method_name class_name method_name in
+     let method_arity = List.length parameters in
+     let overrides = get_overrides env.ie_hierarchy class_name {method_name; method_arity} in
+     let topl_numbering i t = (i, t) in (* topl numbering always from 0 *)
+     let arguments_topl = List.mapi topl_numbering parameters in
+     let return_event_types =
+       if return <> `Void
+       then [(0, as_nonvoid return)]
+       else [] in
+     let call_id = (env.ie_get_tag PA.Call overrides full_method_name arguments_topl) in
+     let return_id = (env.ie_get_tag PA.Return overrides full_method_name return_event_types) in
+     (*let class_type =
+       if virtual_call
+       then [`Class class_name]
+       else [] in
+     let parameters = class_type @ parameters in*)
+     if log_in then printf "\nInvoke Call %s.%s a:%a b:%a\n" (string_of_class_name class_name ) (method_name)
+                    (U.pp_option U.pp_int) call_id (U.pp_option U.pp_int) return_id ;
+     instrument_call locals parameters call_id (instrument_return locals return return_id [instr] )
+                                                             
+let instrument_instruction env locals = function
+  | ( l , BH.INVOKEVIRTUAL (`Methodref ((`Class_or_interface c),m,(parameters,return))) ) as instr ->
+     let parameters = [`Class c] @ parameters in
+     if log_in then printf "\n%Li: InvokeVirtual Call %s.%s \n"
+                    l (string_of_class_name c) (string_of_method_name m) ;
+     instrument_invoke m c parameters return env locals instr
+  | (l, BH.INVOKESTATIC `Methodref((`Class_or_interface c),m,(parameters,return))) as instr ->
+     if log_in then printf "\n%Li: InvokeStatic Call %s.%s \n"
+                    l (string_of_class_name c) (string_of_method_name m) ;
+     instrument_invoke m c parameters return env locals instr
+  | (l, BH.NEW c) as instr ->
+     if log_in then printf "\n%Li: New call for Class %s" l (string_of_class_name c) ;
+     [instr]
```

Comment:
I suppose there is no instrumentation here because each `new` is followed by an `invokespecial` which *is* instrumented. Is that right? If so, can you add a small comment here, please?

Reply:
Yes this is correct, I found it useful to have the debug message there if enabled